validator/db/filesystem/db.go

```diff
@@ -0,0 +1,399 @@
+	s.slashingMuMapMu.Unlock()
+
+	// Read lock the mutex for the public key.
+	mu.RLock()
+	defer mu.RUnlock()
+
+	// Read the file and unmarshal it into ValidatorSlashingProtection struct.
+	yfile, err := os.ReadFile(cleanedPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not read %s", cleanedPath)
+	}
+
+	if err := yaml.Unmarshal(yfile, validatorSlashingProtection); err != nil {
+		return nil, errors.Wrapf(err, "could not unmarshal %s", cleanedPath)
+	}
+
+	return validatorSlashingProtection, nil
+}
+
+// saveValidatorSlashingProtection saves the slashing protection for a public key.
+func (s *Store) saveValidatorSlashingProtection(
+	publicKey [fieldparams.BLSPubkeyLength]byte,
+	validatorSlashingProtection *ValidatorSlashingProtection,
+) error {
+	// If the ValidatorSlashingProtection struct is nil, return.
+	if validatorSlashingProtection == nil {
+		return nil
+	}
+
+	// Create the directory if needed.
+	slashingProtectionDirPath := s.slashingProtectionDirPath()
+	if err := file.MkdirAll(slashingProtectionDirPath); err != nil {
+		return errors.Wrapf(err, "could not create directory %s", s.databasePath)
+	}
+
+	// Get the file path for the public key.
+	path := s.pubkeySlashingProtectionFilePath(publicKey)
+
+	// Lock the mutex protecting the map of public keys to slashing protection mutexes.
+	s.slashingMuMapMu.Lock()
+
+	// Get / create the mutex for the public key.
+	mu, ok := s.pkToSlashingMu[publicKey]
+	if !ok {
+		mu = &sync.RWMutex{}
+		s.pkToSlashingMu[publicKey] = mu
+	}
+
+	// Release the mutex protecting the map of public keys to slashing protection mutexes.
+	s.slashingMuMapMu.Unlock()
+
+	// Write lock the mutex.
+	mu.Lock()
+	defer mu.Unlock()
+
+	// Marshal the ValidatorSlashingProtection struct.
+	yfile, err := yaml.Marshal(validatorSlashingProtection)
+	if err != nil {
+		return errors.Wrap(err, "could not marshal validator slashing protection")
+	}
+
+	// Write the ValidatorSlashingProtection struct to the file.
+	if err := file.WriteFile(path, yfile); err != nil {
+		return errors.Wrapf(err, "could not write into %s.yaml", path)
+	}
+
+	return nil
+}
+
+// publicKeys returns the public keys existing in the database directory.
+func (s *Store) publicKeys() ([][fieldparams.BLSPubkeyLength]byte, error) {
+	// Get the slashing protection directory path.
+	slashingProtectionDirPath := s.slashingProtectionDirPath()
+
+	// If the slashing protection directory does not exist, return an empty slice.
+	if _, err := os.Stat(slashingProtectionDirPath); os.IsNotExist(err) {
+		return nil, nil
+	}
+
+	// Get all entries in the slashing protection directory.
+	entries, err := os.ReadDir(slashingProtectionDirPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not read database directory")
+	}
+
+	// Collect public keys.
+	publicKeys := make([][fieldparams.BLSPubkeyLength]byte, 0, len(entries))
+	for _, entry := range entries {
+		if !(entry.Type().IsRegular() && strings.HasPrefix(entry.Name(), "0x")) {
```

Comment:
Must they be 0x prefixed? Or is it optional in the spec? My worries here is that we just continue and don't log anything to the user. Perhaps we should offer some info that something was ignored and raise a warning

Reply:
There is no spec at all actually for how the data should be stored.
The `0x` prefix is only here to filter potential trash files a user would put in the directory.

> Perhaps we should offer some info that something was ignored and raise a warning.
Definitely, added.